Type variable 2

Yuning30 opened the pull request:
This is the new pull request for type variable, the old type vraiable branch is not used due to conflict with `dev` branch

Commit:
created Kind.re

Commit:
create TyId.re

Commit:
create TyVarCtx.re

Commit:
created TyTextShape.re

Commit:
make the contexts as records and add tyvars field

Commit:
Support TyVar and TyVarHole in UHTyp, update relavent function, create of_string function

Commit:
support fot TyVar and make Int, Float, Bool have text cursrs and update functions accordingly

Commit:
add support to type variables

Commit:
add support for Type variables

Commit:
add TyVar and TyVarHole

Commit:
support type variable and make Int, Float, Bool as text

Commit:
fix spell error for NotInVarHole

Commit:
threads variable ctx for UHTyp.expand

Commit:
add support for type variable

Commit:
add type module and fix UHTyp.expand by feeding ctx

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Commit:
move helper function above module typ

Commit:
add syn_move, mk_syn_result, mk_ana_result, and type syn_success and ana_success

Commit:
make the program conpiles and support new cursorinfo type and support type variable in UndoHistoryPanel and UndoHistory

Commit:
update cell.re to support new CursorInfo.t

Commit:
update cursor info to support new cursorinfo.t

Commit:
make project compile

Commit:
format

Commit:
add ana_move

Commit:
finish mk_ana_text

Commit:
change HTyp.t to 'a to use in module typ

Yuning30 committed:
push for backup

Yuning30 committed:
add syn_cursor_info

Yuning30 committed:
finish cursorinfo

Yuning30 committed:
add ana_perform

Yuning30 committed:
update functions based on TyVarHole behave like Hole

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
support new CursorInfo.OnType

Yuning30 committed:
remove debug print info

Yuning30 commented:
Currently the kind information is displayed via special message bar together with the string "got type", and I'm planning to add a type bar. And you can review the code for other semantics.

cyrus- commented:
OK, I'll take a look soon.

cyrus- reviewed (changes requested):
changes requested verbally

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
add TyVarHole to HTypAnnot.t and DHAnnot.t to describe free type variable. And add .InTyVarHole to style.css

Yuning30 committed:
change error message for type variable to get free type variable

Yuning30 committed:
remove kind related information from Cursorinspector.re until we have more Kind

Yuning30 commented:
#### Changes since last review:
- Now the context inspector and cursor inspector is able to display the red dotted box around the free type variable
- The error message is change to "got a free type variable"
- The kind related information is removed from the cursor inspector until we have more useful kinds

cyrus- commented:
merge conflict needs to be resolved

cyrus- commented:
valid base types (Int, Float, etc.) should display in bold like they do on dev

cyrus- commented on src/hazelcore/Contexts.re:
```diff
@@ -1,19 +1,27 @@
+  tyvars: TyVarCtx.t,
+  palettes: PaletteCtx.t,
+};
+let empty = {
```
leave empty lines between definitions

cyrus- commented on src/hazelcore/CursorInfo.re:
```diff
@@ -95,7 +95,9 @@ type typed =
   | PatSynKeyword(ExpandingKeyword.t)
   /* cursor in type position */
-  | OnType
+  | TypKeyword(ExpandingKeyword.t)
```
what is this for? we don't have any expanding keywords at the type level right?

Yuning30 replied:
I originally thought `let` and `case` are reserved for both expression position and type position, now I'll delete them.

cyrus- replied:
oh I guess we should probably reserve those everywhere for the sake of extraction. you can leave it.

Yuning30 committed:
refmt and make valid base type displayed as bold font

Yuning30 committed:
add empty line between definition and refine tyvar in UndoHistoryPanel

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
add type variables into interface and fix function calls in new dynamics files

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
fix action_typ and statics_typ modules

Yuning30 committed:
finish resolving conflicts

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
fix after merge

Yuning30 commented:
The Tracis-ci seems to fail to update correctly. It's always pending state in the pull request page. But when I click into that page, it says the build is successful.

cyrus- commented on src/hazelcore/UHTyp.re:
```diff
@@ -95,35 +99,41 @@ let contract = (ty: HTyp.t): t => {
-  fun
-  | OpSeq(skel, seq) => expand_skel(skel, seq)
-and expand_skel = (skel, seq) =>
+let rec expand = (ty: t, ctx: TyVarCtx.t): HTyp.t => expand_opseq(ty, ctx)
```
make the ctx the first argument here, to be consistent with the judgement form

cyrus- commented on src/hazelcore/UHTyp.re:
```diff
@@ -149,3 +161,10 @@ and is_complete = (ty: t) => {
   };
 };
+
+let of_string =
```
this should be called "to_string" not "of_string"

cyrus- commented on src/hazelcore/UHTyp.rei:
```diff
@@ -28,17 +29,19 @@ let mk_OpSeq: OpSeq.seq(operand, operator) => OpSeq.t(operand, operator);
 let contract: HTyp.t => t;
 
-let expand: t => HTyp.t;
+let expand: (t, TyVarCtx.t) => HTyp.t;
```
put contexts first in all of these

cyrus- commented on src/hazelcore/ZTyp.re:
```diff
@@ -74,10 +76,11 @@ and is_after_zopseq = zopseq => ZOpSeq.is_after(~is_after_zoperand, zopseq)
-  | CursorT(cursor, Float)
-  | CursorT(cursor, Bool) => cursor == OnDelim(0, After)
+  | CursorT(cursor, Unit) => cursor == OnDelim(0, After)
+  | CursorT(cursor, Int) => cursor == OnText(3)
```
instead of magic numbers here, define constants determined by calling length on the string representation of each type (based on to_string)

cyrus- commented on src/hazelcore/dynamics/Evaluator.re:
```diff
@@ -20,13 +20,15 @@ let grounded_List = NotGroundOrHole(List(Hole));
   | Sum(Hole, Hole)
-  | List(Hole) => Ground
+  | List(Hole)
+  | TyVar(_, _) => Ground
```
TyVar shouldn't be ground, it depends on what it stands for. leave it as a failwith("Unimplemented") for now, since we don't have bound type variables yet.

cyrus- commented on src/hazelcore/layout/UHAnnot.re:
```diff
@@ -9,6 +9,7 @@ type term_data = {
 
 [@deriving sexp]
 type token_shape =
+  | BoldText
```
call this Keyword rather than BoldText

cyrus- commented on src/hazelweb/www/style.css:
```diff
@@ -814,6 +815,9 @@ h1 {
 .code-delim {
   font-weight: bold;
 }
+.code-boldtext {
```
rename to keyword everywhere

cyrus- commented on src/hazelcore/TyId.re:
```diff
@@ -0,0 +1,18 @@
+open Sexplib.Std;
```
include a .rei file for every .re file you add

cyrus- commented on src/hazelcore/Statics_Typ.re:
```diff
@@ -0,0 +1,19 @@
+let rec syn = (ctx: Contexts.t, ty: HTyp.t): Kind.t =>
+  switch (ty) {
+  | TyVar(idx, _) =>
+    let (_, k) = List.nth(ctx.tyvars, idx);
```
Define this as a function TyVarCtx, rather than assuming it will always be a list

cyrus- commented on src/hazelcore/Statics_Typ.re:
```diff
@@ -0,0 +1,19 @@
+  | Prod([]) => Type
+  | Arrow(ty1, ty2)
+  | Sum(ty1, ty2) =>
+    ana(ctx, ty1, Kind.Type) && ana(ctx, ty2, Kind.Type) ? Type : KHole
```
if analysis fails on either argument, then synthesis should fail, not return KHole. need to change return type to option.

cyrus- commented on src/hazelcore/Statics_Typ.re:
```diff
@@ -0,0 +1,19 @@
+  | Arrow(ty1, ty2)
+  | Sum(ty1, ty2) =>
+    ana(ctx, ty1, Kind.Type) && ana(ctx, ty2, Kind.Type) ? Type : KHole
+  | Prod([x, ...xs]) => ana(ctx, x, Kind.Type) ? syn(ctx, Prod(xs)) : KHole
```
need to check each element against Type, not synthesize. and it shouldn't return KHole, it should return Type if this succeeds

cyrus- commented on src/hazelcore/Statics_Typ.re:
```diff
@@ -0,0 +1,19 @@
+  | Sum(ty1, ty2) =>
+    ana(ctx, ty1, Kind.Type) && ana(ctx, ty2, Kind.Type) ? Type : KHole
+  | Prod([x, ...xs]) => ana(ctx, x, Kind.Type) ? syn(ctx, Prod(xs)) : KHole
+  | List(t) => ana(ctx, t, Kind.Type) ? Type : KHole
```
return None if it fails

cyrus- commented on src/hazelcore/HTyp.re:
```diff
@@ -42,13 +49,24 @@ let precedence = (ty: t): int =>
-let eq = (==);
+let eq = (x, y) =>
+  switch (x, y) {
+  | (TyVarHole(_), TyVarHole(_))
```
we shouldn't call all TyVarHoles equal to each other or to Holes. they should only be equal if they are the same variable.

cyrus- commented on src/hazelcore/Action_Typ.re:
```diff
@@ -24,6 +24,27 @@ let shape_of_operator = (op: UHTyp.operator): Action_common.operator_shape =>
   | Sum => SVBar
```
haven't reviewed this file yet, will take a look in next review

cyrus- reviewed (changes requested):
- see comments
- need to add support for splitting, similar to expression and pattern variables (e.g. if I am at `a|b` and type `>`, it should become `a -> b`)

Yuning30 commented:
Thanks for the feedback. I'll fix these issues ASAP.

Yuning30 committed:
change Boldtext to Keyword, eliminate magic numbers and put context the first argument

Yuning30 committed:
fix argument order where UHTyp.expand appears

Yuning30 committed:
change return type of syn to option(Kind.t)

Yuning30 committed:
fix HTyp and all places where Statics_Typ.syn appears

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
fix after merge dev

Yuning30 committed:
implement split text for Typ

Yuning30 committed:
add interface files for Kind, TyId, TytextShape, TyVarCtx

Yuning30 committed:
remove List.nth by using functions in TyCtx modules

cyrus- commented:
I get this failure when trying to enter most forms into hole 5 (e.g. number literals, lambdas)...
![Screenshot from 2020-06-23 15-13-58](https://user-images.githubusercontent.com/280638/85447963-2debb300-b564-11ea-849a-bc57d25a7dbf.png)

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Commit:
fix after merge

Commit:
add TyVarHole to Evaluator, bug fixed

Yuning30 committed:
Merge branch 'dev' into type-variable-2

Yuning30 committed:
fix action panel support for type variable

bkase commented on src/hazelcore/dynamics/Evaluator.re:
```diff
@@ -20,13 +20,15 @@ let grounded_List = NotGroundOrHole(List(Hole));
   | Arrow(Hole, Hole)
   | Sum(Hole, Hole)
   | List(Hole) => Ground
+  | TyVar(_, _) => failwith("Unimplemented")
```
remember to add context for calls to a function

cyrus- commented:
Superceded by #478
